pfcpiface/bess.go

```diff
@@ -677,72 +690,151 @@ func (b *bess) processQER(ctx context.Context, any *anypb.Any, method upfMsgType
+				Cir:  cir, /* committed info rate */
+				Pir:  pir, /* peak info rate */
+				Cbs:  cbs, /* committed burst size */
+				Pbs:  pbs, /* Peak burst size */
+				Ebs:  ebs, /* Excess burst size */
+				Fields: []*pb.FieldData{
+					intEnc(uint64(srcIface)),  /* Src Intf */
+					intEnc(uint64(qer.qerID)), /* qer_id */
+					intEnc(uint64(qer.fseID)), /* fseid */
+				},
+				Values: []*pb.FieldData{
+					intEnc(uint64(qer.qfi)), /* QFI */
+				},
+			}
+			any, err = anypb.New(q)
+			if err != nil {
+				log.Println("Error marshalling the rule", q, err)
+				return
+			}
+			b.processQER(ctx, any, upfMsgTypeAdd)
 		}
 
-		b.processQER(ctx, any, upfMsgTypeAdd)
+		if qer.dlMbr != 0 {
+			pir = (qer.dlMbr * 1000) / 8
+			cir = (qer.dlMbr * 1000) / 8
+			if qer.dlGbr != 0 {
+				cir = (qer.dlGbr * 1000) / 8
+			}
+			if cir >= MaxUint32 {
+				cir = MaxCir
+			}
+			if pir >= MaxUint32 {
+				pir = MaxPir
+			}
+			srcIface = core
+			q := &pb.QosCommandAddArg{
+				Gate: uint64(0),
+				Cir:  cir, /* committed info rate */
+				Pir:  pir, /* peak info rate */
+				Cbs:  cbs, /* committed burst size */
+				Pbs:  pbs, /* Peak burst size */
+				Ebs:  ebs, /* Excess burst size */
+				Fields: []*pb.FieldData{
+					intEnc(uint64(srcIface)),  /* Src Intf */
+					intEnc(uint64(qer.qerID)), /* qer_id */
+					intEnc(uint64(qer.fseID)), /* fseid */
+				},
+				Values: []*pb.FieldData{
+					intEnc(uint64(qer.qfi)), /* QFI */
+				},
+			}
+			any, err = anypb.New(q)
+			if err != nil {
+				log.Println("Error marshalling the rule", q, err)
+				return
+			}
+			b.processQER(ctx, any, upfMsgTypeAdd)
+		}
 		done <- true
 	}()
 }
 
 func (b *bess) delQER(ctx context.Context, done chan<- bool, qer qer) {
 	go func() {
-		var (
-			any *anypb.Any
-			err error
-		)
-
-		q := &pb.ExactMatchCommandDeleteArg{
-			Fields: []*pb.FieldData{
-				intEnc(uint64(qer.qerID)), /* qer_id */
-				intEnc(qer.fseID),         /* fseid */
-			},
+		var any *anypb.Any
+		var err error
+		var srcIface uint8
+		if qer.ulMbr != 0 {
+			srcIface = access
+			q := &pb.QosCommandDeleteArg{
+				Fields: []*pb.FieldData{
+					intEnc(uint64(srcIface)),  /* Src Intf */
+					intEnc(uint64(qer.qerID)), /* qer_id */
+					intEnc(uint64(qer.fseID)), /* fseid */
+				},
+			}
+			any, err = anypb.New(q)
+			if err != nil {
+				log.Println("Error marshalling the rule", q, err)
+				return
+			}
+			b.processQER(ctx, any, upfMsgTypeDel)
 		}
 
-		any, err = anypb.New(q)
-		if err != nil {
-			log.Println("Error marshalling the rule", q, err)
-			return
+		if qer.dlMbr != 0 {
+			srcIface = core
+			q := &pb.QosCommandDeleteArg{
+				Fields: []*pb.FieldData{
+					intEnc(uint64(srcIface)),  /* Src Intf */
+					intEnc(uint64(qer.qerID)), /* qer_id */
+					intEnc(uint64(qer.fseID)), /* fseid */
+				},
+			}
+			any, err = anypb.New(q)
+			if err != nil {
+				log.Println("Error marshalling the rule", q, err)
+				return
+			}
+			b.processQER(ctx, any, upfMsgTypeDel)
```

Comment:
can we replace this will loop ranging over `[]{core, access}` since its just a delete? Could do the same in add?